Refresh COMMAND, add topics: command-args, key-specs (#1697)
redis/redis-doc: topics/command-tips.md

### random-order-output

The output is deterministic, but comes in random order (e.g. `HGETALL`, `SMEMBERS`)

### block_keyword

Identify commands which the potential to block, only is a specific arguement was given: For example, `XREAD` and `XREADGROUP` have the `blocking` tip, but they will only block if `BLOCK` was given. We will add the `block_keyword:<word>` tip, and the client should search `<word>` search in `argv` to determine if the command may indeed block.
This could result in false-positive if that keyword is a keyname or some other form of free-text, but this tip best-effort: there's little harm in mistaking a non-blocking command with a blocking one (but mistaking in the other direction is unacceptable).

### request_policy

This tip should help clients to determine to which shard(s) to send the command in clustering mode.
The default behavior (i.e. if the `request_policy` tip is absent) devides into two cases:
1. The command has key(s). In this case the command goes to a single shard, determined by the hslot(s) of the key(s)

> 2. The command doesn't have key(s). In this case the command can be sent to any arbitrary shard. In order to correctly process certain key-less commands, they should always go to the same shard (Example: Pub/Sub will not work properly if `SUBSCRIBE` and `PUBLISH` would reach different shards).

Reviewer: Pubsub works "correctly" if the commands reach different shards.

Author: i'll remove the last sentence

If the client need to behave differently we must specify an option for `request_policy`:

> - **all_shards** - Forward to all shards (`CONFIG SET`). Usually used on key-less command. The operation is atomic on all shards.

Reviewer: I think we should make the distinction between "all nodes" and "all primaries". `Config SET` should be on all nodes, `FUNCTION CREATE` should be on all primaries. Also the "The operation is atomic on all shards" is confusing.

Reviewer 2: "primaries" it not a term use by redis.. don't wanna add further mixup of terms.
let's use `all_shards` (mention "masters" in the doc), and `all_nodes` (mention that it includes replicas in the doc)

# Command tips

An @array-reply of strings, command tips are meant to provide additional information about commands for clients and proxies.
This infromation should help clients and proxies to determine how to handle this command when Redis is in clustering mode.

> Unlike the command flags (the thirds elemnts in `COMMAND`'s reply), which are internal in nature (Redis itself uses them internally), command tips are just exposed via `COMMAND` and Redis does not use them.

Reviewer 2: ```suggestion
Unlike the command flags (the third element in `COMMAND`'s reply), which are internal in nature (Redis itself uses them internally), command tips are just exposed via `COMMAND` and Redis does not use them.
```

The tips are arbirary strings, but in order to create some sort of concensus, here is a list of suggested tips:

## Boolean tips

Tips that either exist or not, no value is associated with them.

- **blocking**
- **random-output**
- **random-order-output**

### blocking

If present, this command has the potential to block (but not always, see `block_keyword` below)

### random-output

The output of the command is random (e.g. `RANDOMKEY`, `SPOP`)

### random-order-output

The output is deterministic, but comes in random order (e.g. `HGETALL`, `SMEMBERS`)

## Key-value tips

Tips that are associated with a value.

- **block_keyword** 
- **request_policy**
- **reply_policy**

### block_keyword

> Identify commands which the potential to block, only is a specific arguement was given: For example, `XREAD` and `XREADGROUP` have the `vlocking` tip, but they will only block if `BLOCK` was given. We will add the `block_keyword:<word>` tip, and the proxy should search `<word>` search in `argv` to determine if the command may indeed block.

Reviewer 2: ```suggestion
Identify commands which the potential to block, only is a specific arguement was given: For example, `XREAD` and `XREADGROUP` have the `blocking` tip, but they will only block if `BLOCK` was given. We will add the `block_keyword:<word>` tip, and the proxy should search `<word>` search in `argv` to determine if the command may indeed block.
```

This could result in false-positive if that keyword is a keyname or some other form of free-text, but this tip best-effort: there's little harm in mistaking a non-blocking command with a blocking one (but mistaking in the other direction is unacceptable).

### request_policy

This tip should help proxies/client to determine to which shard(s) to send the command.
The default behavior (i.e. if the `request_policy` tip is absent) devides into two cases:
1. The command has key(s). In this case the command goes to a single shard, sdetemined by the hslot(s) of the key(s)
2. The command doesn't have key(s). In this case the command goes to an arbitrary shard (usually the one with the lowest hash slot).

If the proxy/client need to behave differently we must specify an option for `request_policy`:

> - **all_shards** - Forward to all shards (`PING`). Usualt used on key-less command. The operation is atomic on all shards.

Reviewer 2: ```suggestion
- **all_shards** - Forward to all shards (`CONFIG SET`). Usually used on key-less command. The operation is atomic on all shards.
```

> - **few_shards** - Forward to several shards, used by multi-key commands (`MSET`, `MGET`, `DEL`, etc.). The operation is atomic on relevant shards only.

Reviewer 2: ```suggestion
- **few_shards** - Forward to several shards, used by multi-key commands where each key is handled separately (`MSET`, `MGET`, `DEL`, etc.). i.e. unlike `SUNIONSTORE` which must be sent to one shard.
```

### reply_policy

This tip should help clients/proxies to know how to aggregate the replies of a command that was sent to multiple shards.
The default behavior (i.e. if the `request_policy` tip is absent) applies only when the reply is some sort of collection (array, set, map) and it devides into two cases:

> 1. The command has key(s). In this case we append the array replies in random order (e.g. `KEYS`)
> 2. The command doesn't have key(s).  In this case we append the array replies in the origianl order of the request's keys (e.g. `MGET`, but not `MSET` and `DEL` which don't return an array)

Reviewer 2: ```suggestion
1. The command doesn't have key(s). In this case we append the array replies in random order (e.g. `KEYS`)
2. The command has key(s).  In this case we append the array replies in the original order of the request's keys (e.g. `MGET`, but not `MSET` and `DEL` which don't return an array)
```

If the reply is not a collection, or if the proxy/client need to behave differently we must specify an option for `reply_policy`:

> - **all_succeeded** - Return OK if all shards didn't reply with an error. All the replies should be identical and the proxy replies back to client with one of them. Examples: `CONFIG SET`, `SCRIPT FLUSH`, `SCRIPT LOAD`
> - **one_succeeded** - Return OK if at least one shard didn't reply with an error. The proxy should reply with the first reply it gets which isn't an error. Example: `SCRIPT KILL` (usually the script is loaded to all shards, but runs only on one. `SCRIPT KILL` is sent to all shards)

Reviewer 2: ```suggestion
- **one_succeeded** - Return success if at least one shard didn't reply with an error. The client/proxy should reply with the first reply it gets which isn't an error, or with any of the errors, if they all responded with errors. Example: `SCRIPT KILL` (usually the script is loaded to all shards, but runs only on one. `SCRIPT KILL` is sent to all shards)
- **all_succeeded** - Return success if none of the shards replied with an error. if one replied with an error, return that error (either one of the errors), if they all succeeded, return the successful reply (either one). Examples: `CONFIG SET`, `SCRIPT FLUSH`, `SCRIPT LOAD`
```

- **agg_logical_and** - Preform a logical AND on the replies (replies must be numerical, usually just 0/1). Example: `SCRIPT EXISTS`, which returns an array of 0/1 indicating which of the provided scripts exist. The aggregated response will be 1 iff all shards have the script.

> - **agg_logical_and** - Preform a logical OR on the replies (replies must be numerical, usually just 0/1).

Reviewer 2: ```suggestion
- **agg_logical_or** - Preform a logical OR on the replies (replies must be numerical, usually just 0/1).
```

> - **agg_min** - Perform a minimum on replies (replies must be numerical). Example: `WAIT` (returns the lowest number among the ones the shards` replies).

Reviewer 2: ```suggestion
- **agg_min** - Perform a minimum on replies (replies must be numerical). Example: `WAIT` (returns the lowest number among the ones the shards' replies).
```

# Command tips

An @array-reply of strings, command tips are meant to provide additional information about commands for clients and proxies.

> This infromation should help clients and proxies to determine how to handle this command when Redis is in clustering mode.

Reviewer 3: Isn't a proxy just a special case of a client? Not sure if this specific reference helps the reader or just creates confusion.

Reviewer 2: i removed a few more mentions of "proxy", and left only the one at the very top "proxies"

Unlike the command flags (the third element in `COMMAND`'s reply), which are internal in nature (Redis itself uses them internally), command tips are just exposed via `COMMAND` and Redis does not use them.

> The tips are arbirary strings, but in order to create some sort of concensus, here is a list of suggested tips:

Reviewer 3: ```suggestion
The tips are arbitrary strings, but in order to create common conventions, here is a list of suggested tips:
```

> ## Boolean tips

Reviewer 3: I find the grouping of tips by their format/associated value a bit wierd. I think for a user, a single flat table with tips, their format/value and a description will be more usable.

Reviewer 2: one reason why they're grouped is because the request and reply policies have default that defines the behavior of these aspects when no explicit flag is set.

Reviewer 2: I see Guy listed the tips at the top, but not all combinations.
i.e. it lists `request_policy:<policy>` not `request_policy:all_shards`, `request_policy:few_shards`.
i'm ok with that, but i don't think that's what Yossi meant (if i understand what Yossi meant, i think it's problematic)

Tips that either exist or not, no value is associated with them.

- **blocking**
- **random-output**
- **random-order-output**

### blocking

If present, this command has the potential to block (but not always, see `block_keyword` below)

### random-output

The output is not deterministic, that is, the same command with the same arguments, with the same key-space, may have different results. (e.g. `RANDOMKEY`, `SPOP`)

### random-order-output

The output is deterministic, but comes in random order (e.g. `HGETALL`, `SMEMBERS`)

## Key-value tips

Tips that are associated with a value.

- **block_keyword** 
- **request_policy**
- **reply_policy**

### block_keyword

Identify commands which the potential to block, only is a specific arguement was given: For example, `XREAD` and `XREADGROUP` have the `blocking` tip, but they will only block if `BLOCK` was given. We will add the `block_keyword:<word>` tip, and the proxy should search `<word>` search in `argv` to determine if the command may indeed block.

> This could result in false-positive if that keyword is a keyname or some other form of free-text, but this tip best-effort: there's little harm in mistaking a non-blocking command with a blocking one (but mistaking in the other direction is unacceptable).

Reviewer 3: Are we confident there's always little harm in false positive id of a blocking command?

I'm not too sure about that. For example, if a client will use this information to manage connections, such a false positive may lead to a connection pool exhausted unexpectedly only because specific key names / values were used.

Author: good point, but do you have suggestions for improvement? do something like key-specs?
FYI it's only relevant for XREAD and XREADGROUP, where the only opportunity for a rouge "block" is the keyname. maybe it's enough just to mention that we need to search argv from the beginning, ignoring args which are keys?

Reviewer 2: did this design come from the RL proxy?
are there any modules that we're aware of that implement such an argument?
anyway, if we can't agree on it, i'm willing to leave it out of the scope for now, we can always add it later.

unless we wanna give an example of a tip in which part of the tip is a non-const argument for that tip

Reviewer 2: @yossigo it's up to you to decide.
either we keep this flag and document the limitation
i.e. don’y have any non-key free text before the optional blocking keyword.
or we drop it for now, and just mark these commands as potentially blocking..
let me know what you prefer.

Reviewer 2: BTW, this is related to this request: https://github.com/redis/redis/issues/3642

Reviewer 2: we decided to drop it.

Author: removed block_keyword

> ### request_policy

Reviewer 3: Nothing here says `cluster` and I think it should, to set the right context.
And another nit: we always use the term "command", not "request".

Author: "request" is used as a mirror for "reply", I'm open to suggestions... "command" instead of "request", but what will replace "reply"?

Author: (regardless, i will mention the word cluster)

Reviewer 2: i don't think the term "command" is good.
"command" is what's being executed by redis.
i think "request" is the intent of the client app, which may need to be propagated to multiple shards and aggregate their response.
bottom line, i'm ok with the current term. i'm not strictly against "command" but i must say i don't like it.

Reviewer: I like request more than command, but also don't feel that strongly.

Author: will will change reply -> response, keep request

This tip should help proxies/client to determine to which shard(s) to send the command.
The default behavior (i.e. if the `request_policy` tip is absent) devides into two cases:
1. The command has key(s). In this case the command goes to a single shard, determined by the hslot(s) of the key(s)

> 2. The command doesn't have key(s). In this case the command goes to an arbitrary shard (usually the one with the lowest hash slot). Different key-less commands should always go to the same shard.

Reviewer 3: ```suggestion
2. The command doesn't have key(s). In this case the command can be sent to any arbitrary shard. Different key-less commands should always go to the same shard.
```

Can we explain the last sentence and give an example? It's not clear/obvious.

Author: it means that all PIBLISHs and SUBSCRIBEs should all go to the same shard, and not randomly select a shard for each key-less command

- **all_succeeded** - Return success if none of the shards replied with an error. if one replied with an error, return that error (either one of the errors), if they all succeeded, return the successful reply (either one). Examples: `CONFIG SET`, `SCRIPT FLUSH`, `SCRIPT LOAD`
- **agg_logical_and** - Preform a logical AND on the replies (replies must be numerical, usually just 0/1). Example: `SCRIPT EXISTS`, which returns an array of 0/1 indicating which of the provided scripts exist. The aggregated response will be 1 iff all shards have the script.
- **agg_logical_or** - Preform a logical OR on the replies (replies must be numerical, usually just 0/1).
- **agg_min** - Perform a minimum on replies (replies must be numerical). Example: `WAIT` (returns the lowest number among the ones the shards' replies).
- **agg_max** - Perform a maximum on replies (replies must be numerical).
- **agg_sum** - Sums the integer values returned by the shards. Example: `DBSIZE`

> - **non_trivial** - Indicates a non-trivial form of reply policy. Example: `INFO`

Reviewer 3: Can we assume that if no policy is specified, it's by definition "non-trivial"? Applies also for request policy.

Author: we chose the default ones because they are the most common. if non-trivial is the default we will have to mention either FIRST_SHARD or SLOT_SHARD as request_policy in the json files of the vast majority of commands

Author: @oranagra WDYT?

Reviewer 2: i agree with Guy.. i rather mention a sane default, and then flag the few command that can't match it and may need special handling.

Author: how about "special"?

Author: we went with "special"
